File: leetcode/139.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Solution:

    def wordBreak(self, s: str, wordDict: List[str]) -> bool:

        d = set(wordDict)
        nd = set()

        def solve(st):
            if st in nd:
                return False
            if st in d:
                return True

            for word in d:
                if st.startswith(word):
                    if self.solve(st[:len(word)]):
                        return True
                    else:
                        nd.add(st[:len(word)])
            return False

        return solve(s)

File: leetcode/139.cpp
class Solution {
public:
    bool solve(string s, unordered_set<string>& wordDict) {

      int n, i, j;
      string word;
      n = s.length();
      vector<bool> dp(n+1, false);
      dp[0] = true;
      for (i = 0; i <= n; ++i){
        for (j = i-1; j >= 0; --j) {
          if (dp[j]) {
            word = s.substr(j, i-j);
            if (wordDict.find(word) != wordDict.end()) {
              dp[i] = true;
              break;
            }
          }
        }
      }
      return dp[n];
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> d(wordDict.begin(), wordDict.end());
        return solve(s, d);
    }
};
